### PR TITLE
New generic matching mechanism for matching callbacks

### PR DESCRIPTION
**Pull request by DonAurelio**
* We updated callbacks models specifying matching callbacks to a generic model complying the metababel model. 
* We split the mechanism in two phases, matching and extraction of arguments.
* We validate conflicting function signatures for matching callbacks.
* We lost the flexibility to indicate the order of the arguments, then arguments are passed to the callbacks in the same order as they are defined in the model, since they are extracted recursivelly.

### CODE REVIEW

File: bin/metababel [hunk 1]
```diff
@@ -122,44 +103,51 @@ end
 # Indeeed the <%= body %> will be indented, but we don't don't want it,
 # in the body string is already indented
 # But we clean the white space empty line afterward \o/
-def wrote_event_dispatchers(folder, t, matching_dispatch_data, callback_types)
+def wrote_event_dispatchers(folder, trace_model, trace_matcher, callback_types)
```
Comment (TApplencourt) [inline bin/metababel:106]: maybe trace_matching_model ?

File: bin/metababel [hunk 2]
```diff
@@ -122,44 +103,51 @@ end
...
       body, index_stream_class, index_event_class,
       default_clock_class,
       arg_variables.fetch('outputs_allocated', []),
-      dispatch_types
+      [dispatch_type],
+      e
     )
   end
 
+  trace_matcher.map_event_classes do |em|
```
Comment (TApplencourt) [inline bin/metababel:131]: Mayve a little comment
"Matching callbacks. Mutate the dispatchera dispatch_types"

File: lib/metababel/bt2_trace_class_generator.rb [hunk 3]
```diff
@@ -224,15 +242,34 @@ def get_getter(event:, arg_variables:)
...
+        stream_class.default_clock_class ? (not (@default_clock_class.nil? or stream_class.default_clock_class.nil?)) : true ].all?
+    end
+
+    def get_args(stream_class)
+      [ @parent.get_args(stream_class.parent),
+        stream_class.packet_context_field_class ? @packet_context_field_class.get_args(stream_class.packet_context_field_class) : nil,
+        stream_class.event_common_context_field_class ? @event_common_context_field_class.get_args(stream_class.event_common_context_field_class) : nil,
+        # _timestamp should be updated in the upstream.c.erb template if changed.
+        stream_class.default_clock_class ? (stream_class.default_clock_class.fetch(:extract, true) ? GeneratedArg.new('int64_t', '_timestamp') : nil) : nil ].flatten.compact
+    end
```
Comment (TApplencourt) [inline lib/metababel/bt2_trace_class_generator.rb:260]: How we do in normal callback? Why we don't need it when using normal callback?

File: lib/metababel/bt2_trace_class_generator.rb [hunk 4]
```diff
@@ -717,6 +802,22 @@ def get_getter(variable:, arg_variables:)
...
+      end
+    end
+
+    def get_args(field_class)
+      field_class.members.filter { |member| member.extract }.map do |rhs_member| 
+        members_matched = @members.find_all { |member| member.match?(rhs_member) }
+        raise "rhs_member '#{rhs_member}' must match one member '#{ members_matched.length }' matched." unless members_matched.length == 1
+        members_matched
+      end.flatten.map(&:get_arg)
+    end
```
Comment (TApplencourt) [inline lib/metababel/bt2_trace_class_generator.rb:820]: Will it be better if `match` return either False or the list of filter it matches?
s? I may avoid code duplication ( and add just one function not 2 :)

Or just use `get_args` who return empty list if not matching.
What does `extract` doe?

I found this function a little complicated with the filter + extract + flatten. A little comment will be appreciated to explain the goal of this function

File: test/callbacks/cases_matching_callbacks/4.btx_callbacks.yaml [hunk 5]
```diff
@@ -1,4 +1,7 @@
-- :name: usr_event
-  :regex: /event/
-  :args:
-  - _timestamp
+:match: true
```
Comment (TApplencourt) [inline test/callbacks/cases_matching_callbacks/4.btx_callbacks.yaml:1]: What is this `match: true`?

**Comment by DonAurelio** [comment]
Almost done, I need to refactor the code in metababel bin to avoid this dispatcher's reversal.

**Comment by TApplencourt** [comment]
Nice \o/ well done! Will review :)

**Comment by DonAurelio** [comment]
Applied @TApplencourt  suggestion to remove the dispatcher's reversal. I will take a look on everything just to be sure we have what we need.

**Comment by DonAurelio** [comment]
I think we are good to go now.

**Pull request merged**